### PR TITLE
Transceiver Support

### PR DESCRIPTION
**Pull request by murrant**
Support for transceivers and metrics.
Includes support for optics with multiple channels/lanes
API access to transceivers and metrics and to set thresholds
UI on ports page, port sub tab, device overview, and threshold settings
Initial OS support:

- Cisco
- Juniper Junos
- HPE Comware
- Mikrotik RouterOS
- FS S5850 (52642)
- Calix EXA (GPON only)
- IP Infusion OcNOS

*Screenshots may have changed

Ports List:
image

Transceiver Full view:
image

Device Overview:
image

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [x] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by freddy36** [comment]
I very much like this approach of separate handling of transceivers. Especially the sensor type grouping is annoying when doing it via sensors.

Any plans to allow adding mappings via yaml instead of having to write PHP code?

**Comment by murrant** [comment]
@freddy36 I haven't decided if that is in the scope of this PR or not.


If anyone is testing this, I haven't included the compiled javascript/css yet, so you need to run `npm run prod` after applying the patch.

**Comment by freddy36** [comment]
I tested it and in general it's working fine.

For multi channel transceviers the naming looks a little bit strange:
image
image
Would suggest to always include the channel number, so `Rx Power` becomes `Channel 0 Rx Power`
Also the mouseover popup only shows the first channel on multichannel Transceviers.

Besides that 'connector' is inconsitently used, on Ocnos you use it to correctly identify the physical connector. On other implementations it's used to distinguish between SM and MM fiber.

**Comment by murrant** [comment]
@freddy36 good eye, that is one of the pending items I need to resolve.  I thought when I first started channels would be 1-4 and all devices without channels would use 0 for the "channel" to signify no channels.  But some devices 0 index the channels (so the first one starts with 0.  I'll get something figured out :)

**Comment by murrant** [comment]
Oh yeah and connector information is inconsistent among devices.  I'm not sure SM/MM is helpful if we have wavelength.  I could split that out into a different field.

File: app/Models/Transceiver.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        'model', // model number or name
+        'revision', // hardware revision
+        'serial', // serial number
+        'date', // date of manufacture
+        'ddm', // if the module supports DDM or DOM
+        'encoding', // data encoding method
+        'cable', // SM, MM, Copper, etc
+        'distance', // Max distance or measured distance
+        'wavelength', // Middle Wavelength in nanometers
+        'connector', // LC, SM, RJ45, etc physical connector type, See Ocnos for some normalized connectors
```
Comment (Jellyfrog) [inline app/Models/Transceiver.php:29]: Wondering if we should enforce the values, there aren't that many connectors. Makes sorting etc nicer.

Comment (murrant) [reply app/Models/Transceiver.php:29]: I counted about 30 connector types :D

Comment (murrant) [reply app/Models/Transceiver.php:29]: https://en.wikipedia.org/wiki/Optical_fiber_connector

Comment (murrant) [reply app/Models/Transceiver.php:29]: We can probably make them consistent by convention instead of enforcing.

**Comment by freddy36** [comment]
Now that https://github.com/librenms/librenms/pull/16175 has been merged, `FsSwitch.php` should become `FsCentec.php`
And `FS-SIWTCH-V2-MIB` should be removed.

File: app/Models/Transceiver.php (hunk 1 above)
Comment (freddy36) [reply app/Models/Transceiver.php:29]: Would suggest to stick to the lists/naming conventions used in the [SFF-8024 specification](https://members.snia.org/document/dl/26423):
image
And odd version returned by switch vendors should be replaced with the standard naming but I wouldn't enforce specific values to allow easy support of new connectors.

**Comment by murrant** [comment]
@freddy36 sure, I'm just second guessing this entire PR as many devices implement this via ENTITY-MIB and ENTITY-SENSOR-MIB.

**Comment by freddy36** [comment]
> @freddy36 sure, I'm just second guessing this entire PR as many devices implement this via ENTITY-MIB and ENTITY-SENSOR-MIB.

Why would that change a lot?

**Comment by murrant** [comment]
Changing this to draft as I don't think this is the correct approach anymore.  I will probably have to scrap all the backend code.

**Comment by peejaychilds** [comment]
> Changing this to draft as I don't think this is the correct approach anymore. I will probably have to scrap all the backend code.

Oh so sad.   Looks like an awesome idea and looked pretty neat in testing.

**Comment by VVelox** [comment]
> Changing this to draft as I don't think this is the correct approach anymore. I will probably have to scrap all the backend code.

@murrant Chimming in on this, I sort of like this approach. Having one that can handle it on a per OS basis allows that to be done where ENTITY-MIB and ENTITY-SENSOR-MIB are not available or do not provide as useful information.

**Comment by electrocret** [comment]
> > Changing this to draft as I don't think this is the correct approach anymore. I will probably have to scrap all the backend code.
> 
> @murrant Chimming in on this, I sort of like this approach. Having one that can handle it on a per OS basis allows that to be done where ENTITY-MIB and ENTITY-SENSOR-MIB are not available or do not provide as useful information.

I'd like to second this. Not all vendors like to support industry standard MIBs and instead make their own specialty MIBs. Having both options would ensure broader support. We'd just have to figure out how to reconcile which to use based on the OS.

**Comment by murrant** [comment]
Sorry, not merging this until I get the ENTITY-MIB stuff figured out.

**Comment by peelman** [comment]
love the concept here. hoping you can work it out. i'll chip in bits to get Nokia's stuff to fit if you get it to where you are happy with it and can use vendor mibs.

**Comment by peelman** [comment]
@murrant im assuming if you had to do magic for EXA some similar magic will be required for AXOS. i will try to pull in your latest changes and test this week; i need to get some work done in that Nokia PR as well 😶

**Comment by murrant** [comment]
@peelman don't get too anxious.  This was just a rebase as prep work to majorly rework this.

**Comment by peelman** [comment]
> @peelman don't get too anxious.  This was just a rebase as prep work to majorly rework this.

anxious is my default state these days. but i'll do my best :)

**Comment by peelman** [comment]
@murrant when you say sensor based metrics will be removed, you mean for OS's that get this new-hotness right? Things that work today will continue to work?  What are you doing about historical data that's already captured via sesnor-based metrics?

**Comment by murrant** [comment]
@peelman I mean I'm converting them to sensors.  This is how most are stored today.  But I will be changing how they are displayed in LibreNMS.

File: misc/alert_rules.json [hunk 2]
```diff
@@ -334,6 +334,14 @@
...
     "name": "Generic PSU status failed"
   },
+  {
+    "builder": {"condition":"AND","rules":[{"id":"transceiver_metrics.status","field":"transceiver_metrics.status","type":"string","input":"text","operator":"greater_or_equal","value":"4"},{"id":"transceiver_metrics.status","field":"transceiver_metrics.status","type":"string","input":"text","operator":"less_or_equal","value":"5"}],"valid":true},
```
Comment (laf) [inline misc/alert_rules.json:338]: transceiver_metrics.status Doesn't exist.

**Review by laf — COMMENTED** [review]
On this page, I'd say either those 4 graphs should all be horizontally aligned or we should place the standard day, week, month, year graphs for each to fill the empty space.
image

**Comment by murrant** [comment]
> On this page, I'd say either those 4 graphs should all be horizontally aligned or we should place the standard day, week, month, year graphs for each to fill the empty space.

For some reason, the CSS didn't automatically update.  Try Ctrl-F5 on the page.  It will look a lot nicer.

**Pull request merged**